### ISSUE TITLE
`launch.py` fails to upload all results data

### ISSUE THREAD

**Issue by nick-pestell** [APHA-CSU/sequence-manager#20]
Getting a "permision" errorr `PermissionError: [Errno 13] Permission denied: 'AFT3-61-03740-21.bam'`. I believe this is only occurring on occasional files: it seems to be failing on bam files for both batches set to the "test" `job_id`.

Weirdly this error doesn't happen if, rather than saving the reads and results in a tmp directory set by python, I choose a permanent hardcoded directory. I have also tried setting the temp directory to be in home rather than at `/tmp/` but I'm still getting the permission error.

```
aws s3 cp --recursive /tmp/tmpriqglt73/results/ s3://s3-staging-area/nickpestell/results//SB4030/NB501786_0379//
upload: ../../../tmp/tmpriqglt73/results/Results_reads_05May22/NonBovID/AFT3-61-03740-21_Contaminated_brackensort.tab to s3://s3-staging-area/nickpestell/results//SB4030/NB501786_0379//Results_reads_05May22/NonBovID/AFT3-61-03740-21_Contaminated_brackensort.tab
upload: ../../../tmp/tmpriqglt73/results/Results_reads_05May22/NonBovID/AFT-61-03769-21_InsufficientData_brackensort.tab to s3://s3-staging-area/nickpestell/results//SB4030/NB501786_0379//Results_reads_05May22/NonBovID/AFT-61-03769-21_InsufficientData_brackensort.tab
upload: ../../../tmp/tmpriqglt73/results/Results_reads_05May22/NonBovID/AFT3-61-03740-21_Contaminated-S1_brackensort.tab to s3://s3-staging-area/nickpestell/results//SB4030/NB501786_0379//Results_reads_05May22/NonBovID/AFT3-61-03740-21_Contaminated-S1_brackensort.tab
upload: ../../../tmp/tmpriqglt73/results/Results_reads_05May22/NonBovID/AFT-61-03769-21_InsufficientData-S1_brackensort.tab to s3://s3-staging-area/nickpestell/results//SB4030/NB501786_0379//Results_reads_05May22/NonBovID/AFT-61-03769-21_InsufficientData-S1_brackensort.tab
upload: ../../../tmp/tmpriqglt73/results/Results_reads_05May22/reads_AssignedWGSCluster_05May22.csv to s3://s3-staging-area/nickpestell/results//SB4030/NB501786_0379//Results_reads_05May22/reads_AssignedWGSCluster_05May22.csv
upload: ../../../tmp/tmpriqglt73/results/Results_reads_05May22/bam/AFT-61-03769-21.bam to s3://s3-staging-area/nickpestell/results//SB4030/NB501786_0379//Results_reads_05May22/bam/AFT-61-03769-21.bam
upload: ../../../tmp/tmpriqglt73/results/Results_reads_05May22/filteredBcf/AFT-61-03769-21_filtered.bcf.csi to s3://s3-staging-area/nickpestell/results//SB4030/NB501786_0379//Results_reads_05May22/filteredBcf/AFT-61-03769-21_filtered.bcf.csi
upload: ../../../tmp/tmpriqglt73/results/Results_reads_05May22/NonBovID/AFT-61-03769-21_InsufficientData_kraken2.tab to s3://s3-staging-area/nickpestell/results//SB4030/NB501786_0379//Results_reads_05May22/NonBovID/AFT-61-03769-21_InsufficientData_kraken2.tab
upload: ../../../tmp/tmpriqglt73/results/Results_reads_05May22/filteredBcf/AFT-61-03769-21_filtered.bcf to s3://s3-staging-area/nickpestell/results//SB4030/NB501786_0379//Results_reads_05May22/filteredBcf/AFT-61-03769-21_filtered.bcf
upload: ../../../tmp/tmpriqglt73/results/Results_reads_05May22/filteredBcf/AFT3-61-03740-21_filtered.bcf to s3://s3-staging-area/nickpestell/results//SB4030/NB501786_0379//Results_reads_05May22/filteredBcf/AFT3-61-03740-21_filtered.bcf
upload: ../../../tmp/tmpriqglt73/results/Results_reads_05May22/reads_BovPos_05May22.csv to s3://s3-staging-area/nickpestell/results//SB4030/NB501786_0379//Results_reads_05May22/reads_BovPos_05May22.csv
upload: ../../../tmp/tmpriqglt73/results/Results_reads_05May22/filteredBcf/AFT3-61-03740-21_filtered.bcf.csi to s3://s3-staging-area/nickpestell/results//SB4030/NB501786_0379//Results_reads_05May22/filteredBcf/AFT3-61-03740-21_filtered.bcf.csi
upload: ../../../tmp/tmpriqglt73/results/Results_reads_05May22/reads_FinalOut_05May22.csv to s3://s3-staging-area/nickpestell/results//SB4030/NB501786_0379//Results_reads_05May22/reads_FinalOut_05May22.csv
upload: ../../../tmp/tmpriqglt73/results/Results_reads_05May22/NonBovID/AFT3-61-03740-21_Contaminated_kraken2.tab to s3://s3-staging-area/nickpestell/results//SB4030/NB501786_0379//Results_reads_05May22/NonBovID/AFT3-61-03740-21_Contaminated_kraken2.tab
upload: ../../../tmp/tmpriqglt73/results/Results_reads_05May22/snpTables/AFT-61-03769-21_snps.tab to s3://s3-staging-area/nickpestell/results//SB4030/NB501786_0379//Results_reads_05May22/snpTables/AFT-61-03769-21_snps.tab
upload: ../../../tmp/tmpriqglt73/results/Results_reads_05May22/snpTables/AFT3-61-03740-21_snps.tab to s3://s3-staging-area/nickpestell/results//SB4030/NB501786_0379//Results_reads_05May22/snpTables/AFT3-61-03740-21_snps.tab
upload: ../../../tmp/tmpriqglt73/results/Results_reads_05May22/vcf/AFT-61-03769-21.vcf.gz to s3://s3-staging-area/nickpestell/results//SB4030/NB501786_0379//Results_reads_05May22/vcf/AFT-61-03769-21.vcf.gz
upload: ../../../tmp/tmpriqglt73/results/Results_reads_05May22/vcf/AFT3-61-03740-21.vcf.gz to s3://s3-staging-area/nickpestell/results//SB4030/NB501786_0379//Results_reads_05May22/vcf/AFT3-61-03740-21.vcf.gz
upload: ../../../tmp/tmpriqglt73/results/Results_reads_05May22/consensus/AFT-61-03769-21_consensus.fas to s3://s3-staging-area/nickpestell/results//SB4030/NB501786_0379//Results_reads_05May22/consensus/AFT-61-03769-21_consensus.fas
upload: ../../../tmp/tmpriqglt73/results/Results_reads_05May22/consensus/AFT3-61-03740-21_unmasked_consensus.fas to s3://s3-staging-area/nickpestell/results//SB4030/NB501786_0379//Results_reads_05May22/consensus/AFT3-61-03740-21_unmasked_consensus.fas
upload: ../../../tmp/tmpriqglt73/results/Results_reads_05May22/bam/AFT3-61-03740-21.bam to s3://s3-staging-area/nickpestell/results//SB4030/NB501786_0379//Results_reads_05May22/bam/AFT3-61-03740-21.bam
upload: ../../../tmp/tmpriqglt73/results/Results_reads_05May22/consensus/AFT3-61-03740-21_consensus.fas to s3://s3-staging-area/nickpestell/results//SB4030/NB501786_0379//Results_reads_05May22/consensus/AFT3-61-03740-21_consensus.fas
upload: ../../../tmp/tmpriqglt73/results/Results_reads_05May22/consensus/AFT-61-03769-21_unmasked_consensus.fas to s3://s3-staging-area/nickpestell/results//SB4030/NB501786_0379//Results_reads_05May22/consensus/AFT-61-03769-21_unmasked_consensus.fas
upload: ../../../tmp/tmpriqglt73/results/report.html to s3://s3-staging-area/nickpestell/results//SB4030/NB501786_0379//report.html
2022-05-05 17:29:41 - [Errno 1] Operation not permitted: '/tmp/tmpriqglt73/results/Results_reads_05May22/bam'
Traceback (most recent call last):
  File "/usr/lib/python3.8/shutil.py", line 673, in _rmtree_safe_fd
    os.unlink(entry.name, dir_fd=topfd)
PermissionError: [Errno 13] Permission denied: 'AFT3-61-03740-21.bam'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "launch.py", line 42, in launch
    run_pipeline_s3(reads_uri, results_uri)
  File "launch.py", line 75, in run_pipeline_s3
    subprocess.run(["aws", "s3", "cp", "--recursive", temp_results, results_uri], check=True)
  File "/usr/lib/python3.8/tempfile.py", line 966, in __exit__
    self.cleanup()
  File "/usr/lib/python3.8/tempfile.py", line 970, in cleanup
    self._rmtree(self.name)
  File "/usr/lib/python3.8/tempfile.py", line 952, in _rmtree
    _rmtree(name, onerror=onerror)
  File "/usr/lib/python3.8/shutil.py", line 718, in rmtree
    _rmtree_safe_fd(fd, path, onerror)
  File "/usr/lib/python3.8/shutil.py", line 655, in _rmtree_safe_fd
    _rmtree_safe_fd(dirfd, fullname, onerror)
  File "/usr/lib/python3.8/shutil.py", line 655, in _rmtree_safe_fd
    _rmtree_safe_fd(dirfd, fullname, onerror)
  File "/usr/lib/python3.8/shutil.py", line 655, in _rmtree_safe_fd
    _rmtree_safe_fd(dirfd, fullname, onerror)
  File "/usr/lib/python3.8/shutil.py", line 675, in _rmtree_safe_fd
    onerror(os.unlink, fullname, sys.exc_info())
  File "/usr/lib/python3.8/tempfile.py", line 937, in onerror
    resetperms(_os.path.dirname(path))
  File "/usr/lib/python3.8/tempfile.py", line 933, in resetperms
    _os.chmod(path, 0o700)
PermissionError: [Errno 1] Operation not permitted: '/tmp/tmpriqglt73/results/Results_reads_05May22/bam'
```

**Comment by afishman** [comment]
did you try running with `sudo python.py launch test`? 

I notice `launch.bash` runs screen with sudo, but I don't think the session it spawns has root privs. I'm wondering if changing line 14 to
```
screen -m bash -c "su; bash install.bash; python3 launch.py $JOB_ID; exec sh"
```
might help.

**Comment by nick-pestell** [comment]
When running `sudo python launch.py test` I get `sudo: python: command not found`

When changing to:  `screen -m bash -c "su; bash install.bash; python3 launch.py $JOB_ID; exec sh"`, I get the following in the new screen session.

```
su: user bash does not exist
usage: launch.py [-h] job
launch.py: error: the following arguments are required: job
```

**Comment by nick-pestell** [comment]
When running `sudo python launch.py test` I get `sudo: python: command not found`

When changing to:  `screen -m bash -c "su; bash install.bash; python3 launch.py $JOB_ID; exec sh"`, I get the following in the new screen session.
```
Password: 
su: Authentication failure
```
after typing my user password.

I also tried with 
`screen -m bash -c "su; bash install.bash; python3 launch.py $JOB_ID; exec bash"`
and got:
```
su: user bash does not exist
usage: launch.py [-h] job
launch.py: error: the following arguments are required: job
```

**Comment by afishman** [comment]
Hmm looks l ok ke the $JOB_ID didn't evaluate

**Comment by afishman** [comment]
Also some error swallowing going on

**Comment by nick-pestell** [comment]
Ok, `sudo python3 launch.py test` fixes it. I forgot to use `python3` before. 

It's weird that the majority of the s3 upload works but fails on a single file...

You said that `launch.bash` runs `screen` with `sudo`. I don't actually see this (there's no `sudo` before `screen`). How did you come to this conclusion @afishman ?

**Comment by afishman** [comment]
you're right, must have misread, sorry!

**Comment by nick-pestell** [comment]
Also, the problem is still there when just running `python3 launch.py test`, so I assume it's unrelated to `screen`. 

Anyway, quick fix would be the change line L14 of `launch.bash` to:
`screen -m bash -c "bash install.bash; sudo python3 launch.py $JOB_ID; exec sh"` See PR #22

**Comment by nick-pestell** [comment]
Fixed: see PR #22
